include/pybind11/cast.h

```diff
@@ -1629,7 +1629,25 @@ struct pyobject_caster {
     template <typename T = type, enable_if_t<std::is_same<T, handle>::value, int> = 0>
     bool load(handle src, bool /* convert */) { value = src; return static_cast<bool>(value); }
 
+#ifdef PYBIND11_DISABLE_IMPLICIT_STR_FROM_BYTES
     template <typename T = type, enable_if_t<std::is_base_of<object, T>::value, int> = 0>
+#else
+    template <typename T = type, enable_if_t<std::is_same<T, str>::value, int> = 0>
+    bool load(handle src, bool /* convert */) {
+        if (isinstance<bytes>(src)) {
+            PyObject *str_from_bytes = PyUnicode_FromEncodedObject(src.ptr(), "utf-8", nullptr);
+            if (!str_from_bytes) throw error_already_set();
+            value = reinterpret_steal<type>(str_from_bytes);
+            return true;
+        }
+        if (!isinstance<type>(src))
+            return false;
+        value = reinterpret_borrow<type>(src);
+        return true;
+    }
+
+    template <typename T = type, enable_if_t<std::is_base_of<object, T>::value && !std::is_same<T, str>::value, int> = 0>
```

Comment:
OK, wow, apparently `py::str(handle)` doesn't do the same as `py::str(bytes)`!??

Anyway, like THIS, then:

```c++
    template <typename T = type, enable_if_t<std::is_base_of<object, T>::value, int> = 0>
    bool load(handle src, bool /* convert */) {
#ifndef PYBIND11_DISABLE_IMPLICIT_STR_FROM_BYTES
        if (std::is_same<T, str>::value && isinstance<bytes>(src)) {
            value = str(src);
            return true;
        }
#endif
        if (!isinstance<type>(src))
            return false;
        value = reinterpret_borrow<type>(src);
        return true;
    }
```

Reply:
Oh, thanks!! The `std::is_same` inside the `load` function is much better! (I was trapped in not considering combining the compile-time condition with the runtime condition.)

`value = str(src)` doesn't work with Python 2 or Python 3, but putting my original code there does the trick.

For completeness, with `value = str(src)` the errors are:

Python2:
```
>           m.pass_to_pybind11_str(malformed_utf8)
E           RuntimeError: Unable to compute length of object
```

Python 3:
```
>       assert m.pass_to_pybind11_str(b"Bytes") == 5  # implicit decode
E       AssertionError: assert 8 == 5
E        +  where 8 = <built-in method pass_to_pybind11_str of PyCapsule object at 0x7f925c32d090>(b'Bytes')
E        +    where <built-in method pass_to_pybind11_str of PyCapsule object at 0x7f925c32d090> = m.pass_to_pybind11_str
```

Reply:
OK, I already get where the Python 3 error originates! In Python 3, `str(b"Bytes") results in `b"Bytes"`!
Trying to understand the Python 2 error, but then there's maybe no way to avoid the C API call code duplication :-(

Reply:
Aha. That's because the unexpected Python 2 error is a bug showing itself. I'll create a PR ;-)

Reply:
There we go: #2392

Reply:
(Can't believe I'm still fixing Python 2 bugs ... :-| )